modules/opentelemetry-common/src/main/scala/io/janstenpickle/trace4cats/opentelemetry/common/OpenTelemetryGrpcSpanExporter.scala

```diff
@@ -24,40 +22,32 @@ object OpenTelemetryGrpcSpanExporter {
     override def getMessage: String = s"Failed to export Open Telemetry span batch to $host:$port"
   }
 
-  def apply[F[_]: Async, G[_]: Foldable](
+  def apply[F[_]: Sync: LiftCompletableResultCode, G[_]: Foldable](
     host: String,
     port: Int,
     makeExporter: ManagedChannel => OTSpanExporter
   ): Resource[F, SpanExporter[F, G]] = {
-    def handleResult(onFailure: => Throwable)(code: CompletableResultCode): F[Unit] =
-      Async[F].async_[Unit] { cb =>
-        val complete = new Runnable {
-          override def run(): Unit =
-            if (code.isSuccess) cb(Right(()))
-            else cb(Left(onFailure))
-        }
-        val _ = code.whenComplete(complete)
-      }
-
     for {
       channel <- Resource.make[F, ManagedChannel](
         Sync[F].delay(ManagedChannelBuilder.forAddress(host, port).usePlaintext().build())
       )(channel => Sync[F].delay(channel.shutdown()).void)
       exporter <- Resource.make(Sync[F].delay(makeExporter(channel)))(exporter =>
-        Sync[F].delay(exporter.shutdown()).flatMap(handleResult(ShutdownFailure(host, port)))
+        Sync[F].delay(exporter.shutdown()).liftResultCode(ShutdownFailure(host, port))
       )
     } yield new SpanExporter[F, G] {
       override def exportBatch(batch: Batch[G]): F[Unit] =
-        handleResult(ExportFailure(host, port))(
-          exporter
-            .`export`(
-              batch.spans
-                .foldLeft(ListBuffer.empty[SpanData]) { (buf, span) =>
-                  buf += Trace4CatsSpanData(Trace4CatsResource(span.serviceName), span)
-                }
-                .asJavaCollection
-            )
-        )
+        Sync[F]
```

Comment:
@janstenpickle please look at this commit carefully

I think there was a bug. `exporter.export(...)` was called eagerly without wrapping in `Sync[F].delay`
This might lead to earlier execution.

I refactored interop with `CompletableResultCode` and make a special typeclass for it (inspired by `monix.catnap.FutureLift`)

This should be backported to `master`. Maybe it is somehow related to the runtime failures I got with this exporter.